👑 feat: Add OIDC Claim-Based Admin Role Assignment

jpsfs opened the pull request:
# Pull Request Template

## Summary

This Pull Request implements #9167 .
The goal is to allow a user that logins OIDC to be an admin in LibreChat depending on a claim that can exist in either the access_token, id_token or it's user profile.

A PR for the documentation was also open, detailing 3 new env vars, similar to the ones already used to check required roles.

```env
OPENID_ADMIN_ROLE=
OPENID_ADMIN_ROLE_PARAMETER_PATH=
OPENID_ADMIN_ROLE_TOKEN_KIND=
```

## Change Type

Please delete any irrelevant options.

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

## Testing

Depending on your OIDC provider configurations might be different.
If you have a claim in the `id_token` called `roles` that's an array, and an admin user must have a role called __admin__, the configuration would be the following:

```env
OPENID_ADMIN_ROLE=admin
OPENID_ADMIN_ROLE_PARAMETER_PATH=roles
OPENID_ADMIN_ROLE_TOKEN_KIND=id
```

### **Test Configuration**:

## Checklist

Please delete any irrelevant options.

- [X] My code adheres to this project's style guidelines
- [X] I have performed a self-review of my own code
- [X] I have commented in any complex areas of my code
- [X] I have made pertinent documentation changes
- [X] My changes do not introduce new warnings
- [X] Local unit tests pass with my changes
- [X] A pull request for updating the documentation has been submitted. https://github.com/LibreChat-AI/librechat.ai/pull/398

jpsfs committed:
feat: Add support for users to be admins when logging in using OpenID

jpsfs committed:
fix: Linting issues

jpsfs committed:
fix: whitespace

jpsfs committed:
chore: add unit tests for OIDC_ADMIN_ROLE

jpaodev commented:
Very cool, I think this makes a lot of sense as alternative and would be a neat upgrade from the current version (first user receives update)

jpsfs commented:
Hi @danny-avila !
Are you open to merging this one? Just need this one to close my fork for now :)

jpsfs commented:
Hi @danny-avila, are you open to merge this?

danny-avila commented:
Hi @jpsfs will be merging this soon, thanks for your patience

danny-avila committed:
refactor: Replace custom property retrieval function with lodash's get for improved readability and maintainability

danny-avila committed:
feat: Enhance OpenID role extraction and error handling in setupOpenId function

- Improved role validation to check for both array and string types.
- Added detailed error messages for missing or invalid role paths in tokens.
- Expanded unit tests to cover various scenarios for nested role extraction and error handling.

danny-avila committed:
fix: Improve error handling for role extraction in OpenID strategy

- Enhanced validation to check for invalid role types (array or string).
- Updated error messages for clarity when roles are missing or of incorrect type.
- Added unit tests to cover scenarios where roles return invalid types (object, number).

danny-avila committed:
feat: Implement user role demotion in OpenID strategy when admin role is absent from token

- Added logic to demote users from 'ADMIN' to 'USER' if the admin role is not present in the token.
- Enhanced logging to capture role changes for better traceability.
- Introduced unit tests to verify the demotion behavior and ensure correct handling when admin role environment variables are not configured.
